src/stores/messages/MessagesDataProviderStore.ts

```diff
@@ -328,37 +321,22 @@ export default class MessagesDataProviderStore implements MessagesDataStore {
 	public onNextChannelResponse = (messages: EventMessage[]) => {
 		this.lastNextChannelResponseTimestamp = null;
 
-		// eslint-disable-next-line no-param-reassign
-		messages = messages.filter(
-			msg => msg.messageId !== this.messagesStore.selectedMessageId?.valueOf(),
-		);
-
-		const prevMessages =
-			this.messages.length > 0
-				? messages
-						.filter(
-							message =>
-								timestampToNumber(message.timestamp) <
-								timestampToNumber(this.messages[0].timestamp),
-						)
-						.sort((a, b) => timestampToNumber(b.timestamp) - timestampToNumber(a.timestamp))
-				: [];
-
 		const nextMessages = messages
-			.slice(0, messages.length - prevMessages.length)
+			.filter(message => !this.messages.map(msg => msg.messageId).includes(message.messageId))
 			.sort((a, b) => timestampToNumber(b.timestamp) - timestampToNumber(a.timestamp));
 
-		if (prevMessages.length > 0 || nextMessages.length > 0) {
-			this.startIndex -= this.updateStore.isActive ? 0 : nextMessages.length;
+		this.startIndex -= this.updateStore.isActive ? 0 : nextMessages.length;
 
-			let newMessagesList = prevMessages.length
-				? [...nextMessages, this.messages[0], ...prevMessages, ...this.messages.slice(1)]
-				: [...nextMessages, ...this.messages];
+		if (nextMessages.length) {
+			let newMessagesList = [...nextMessages, ...this.messages];
+
+			this.startIndex -= this.updateStore.isActive ? 0 : nextMessages.length;
 
 			if (newMessagesList.length > this.messagesLimit) {
-				newMessagesList = newMessagesList.slice(0, this.messagesLimit);
+				newMessagesList = newMessagesList.slice(-this.messagesLimit);
```

Comment:
i think it's source of a bug and we should delete it.
because loaders keep last messageIds, and if you slice some messages loading will start with incorrect messageIds since there is no logic to refetch them